File: frontend/src/Plugins/UserService/APIs/LogUserOperationMessage.ts
/**
 * LogUserOperationMessage
 * desc: 记录用户的操作行为到用户行为日志中。
 * @param userToken: String (用户的身份令牌，用于验证其身份。)
 * @param actionType: String (用户行为的类型，如登录、登出等。)
 * @param actionDetail: String (详细描述用户操作行为的信息。)
 * @return result: String (结果提示字符串，例如'操作记录成功！'。)
 */
import { TongWenMessage } from 'Plugins/TongWenAPI/TongWenMessage'



export class LogUserOperationMessage extends TongWenMessage {
    constructor(
        public  userToken: string,
        public  actionType: string,
        public  actionDetail: string
    ) {
        super()
    }
    getAddress(): string {
        return "127.0.0.1:10010"
    }
}

File: UserService/src/main/scala/APIs/UserService/LogUserOperationMessage.scala
package APIs.UserService

import Common.API.API
import Global.ServiceCenter.UserServiceCode

import io.circe.{Decoder, Encoder, Json}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import io.circe.syntax.*
import io.circe.parser.*
import Common.Serialize.CustomColumnTypes.{decodeDateTime,encodeDateTime}

import com.fasterxml.jackson.core.`type`.TypeReference
import Common.Serialize.JacksonSerializeUtils

import scala.util.Try

import org.joda.time.DateTime
import java.util.UUID


/**
 * LogUserOperationMessage
 * desc: 记录用户的操作行为到用户行为日志中。
 * @param userToken: String (用户的身份令牌，用于验证其身份。)
 * @param actionType: String (用户行为的类型，如登录、登出等。)
 * @param actionDetail: String (详细描述用户操作行为的信息。)
 * @return result: String (结果提示字符串，例如'操作记录成功！'。)
 */

case class LogUserOperationMessage(
  userToken: String,
  actionType: String,
  actionDetail: String
) extends API[String](UserServiceCode)



case object LogUserOperationMessage{
    
  import Common.Serialize.CustomColumnTypes.{decodeDateTime,encodeDateTime}

  // Circe 默认的 Encoder 和 Decoder
  private val circeEncoder: Encoder[LogUserOperationMessage] = deriveEncoder
  private val circeDecoder: Decoder[LogUserOperationMessage] = deriveDecoder

  // Jackson 对应的 Encoder 和 Decoder
  private val jacksonEncoder: Encoder[LogUserOperationMessage] = Encoder.instance { currentObj =>
    Json.fromString(JacksonSerializeUtils.serialize(currentObj))
  }

  private val jacksonDecoder: Decoder[LogUserOperationMessage] = Decoder.instance { cursor =>
    try { Right(JacksonSerializeUtils.deserialize(cursor.value.noSpaces, new TypeReference[LogUserOperationMessage]() {})) } 
    catch { case e: Throwable => Left(io.circe.DecodingFailure(e.getMessage, cursor.history)) }
  }
  
  // Circe + Jackson 兜底的 Encoder
  given logUserOperationMessageEncoder: Encoder[LogUserOperationMessage] = Encoder.instance { config =>
    Try(circeEncoder(config)).getOrElse(jacksonEncoder(config))
  }

  // Circe + Jackson 兜底的 Decoder
  given logUserOperationMessageDecoder: Decoder[LogUserOperationMessage] = Decoder.instance { cursor =>
    circeDecoder.tryDecode(cursor).orElse(jacksonDecoder.tryDecode(cursor))
  }


}

